server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -1336,14 +1336,26 @@ private ClusterState openIndices(final Index[] indices, final ClusterState curre
                     } catch (Exception e) {
                         throw new ElasticsearchException("Failed to verify index " + index, e);
                     }
-                    metadata.getProject(projectId).put(newIndexMetadata, true);
+                    var projectBuilder = projectBuilders.computeIfAbsent(
+                        projectId,
+                        k -> ProjectMetadata.builder(currentState.metadata().getProject(projectId))
+                    );
+                    projectBuilder.put(newIndexMetadata, true);
                 }
 
                 // Always removes index closed blocks (note: this can fail on-going close index actions)
                 blocks.removeIndexBlockWithId(projectId, index.getName(), INDEX_CLOSED_BLOCK_ID);
             }
 
-            ClusterState updatedState = ClusterState.builder(currentState).metadata(metadata).blocks(blocks).build();
+            final Metadata updatedMetadata;
+            if (projectBuilders.isEmpty()) {
+                updatedMetadata = currentState.metadata();
+            } else {
+                final Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
+                projectBuilders.values().forEach(pb -> metadataBuilder.put(pb.build()));
+                updatedMetadata = metadataBuilder.build();
+            }
```

Comment:
I wonder whether this is worth a method on `Metadata`?

Reply:
I thought about that too, but I decided not to do that as this would be the only place that would be able to use the method right now, and I personally don't know any other places that might be able to use that method. I'm fine with moving it to a method if you prefer that, I just don't see a good reason to do so yet.

Reply:
I thought other places have similar patterns. But maybe they are not similar enough to extract a reusable method.